scheduler/src/cook/mesos/sandbox.clj

```diff
@@ -150,19 +136,38 @@
           (contains? #{:success :threshold} reason)
           (update :host->consecutive-failures dissoc hostname)))
 
+(defn- filter-task-ids-with-sandbox
+  "Filters the task-ids which have a sandbox directory from task-id->sandbox."
+  [datomic-conn task-id->sandbox]
+  (letfn [(retrieve-task-ids-with-sandbox [datomic-db task-ids]
+            (d/q '[:find [?t ...]
+                   :in $ [?t ...]
+                   :where
+                   [?e :instance/task-id ?t]
+                   [?e :instance/sandbox-directory ?s]]
+                 datomic-db task-ids))
+          (remove-task-id-with-sandbox [task-id->sandbox task-id]
+            (dissoc! task-id->sandbox task-id))]
+    (->> (keys task-id->sandbox)
+         (retrieve-task-ids-with-sandbox (d/db datomic-conn))
+         (reduce remove-task-id-with-sandbox (transient task-id->sandbox))
+         persistent!)))
+
 (defn refresh-agent-cache-entry
   "If the entry for the specified agent is not cached:
    - Triggers building an indexed version of all task-id to sandbox directory for the specified agent;
-   - After the indexed version is built, it is synced into the task-id->sandbox-agent.
+   - After the indexed version is built, the tasks which have sandbox directories are removed;
+   - The filtered version of tasks without sandbox directories is then synced into the task-id->sandbox-agent.
    The function call is a no-op if the specified agent already exists in the cache."
-  [framework-id mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent agent-hostname]
+  [{:keys [datomic-conn mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent]} framework-id agent-hostname]
   (try
     (let [run (delay
                 (try
-                  (let [task-id->sandbox-directory
-                        (retrieve-sandbox-directories-on-agent framework-id agent-hostname)]
-                    (log/info "Found sandbox directories for" (count task-id->sandbox-directory) "tasks on" agent-hostname)
-                    (send task-id->sandbox-agent aggregate-sandbox task-id->sandbox-directory)
+                  (let [task-id->sandbox-directory (retrieve-sandbox-directories-on-agent framework-id agent-hostname)
```

Comment:
nitpick: Consider this rewrite:

```
(let [task-id->sandbox-directory (->> (retrieve-sandbox-directories-on-agent framework-id agent-hostname)
                                      (filter-task-ids-with-sandbox datomic-conn))] 
   ...)

Reply:
I use `task-id->sandbox-directory` in the logs below to get a sense of items that were dropped.